TEMPLATES YAAAY (#2003)
codesandbox/codesandbox-client: packages/homepage/content/docs/6-templates.md

---
title: 'Templates'
authors: ['Saeris']

> description: 'CodeSandbox allows you to create your  custom templates from sandboxes'

Reviewer: ```suggestion
description: 'CodeSandbox allows you to create custom templates from sandboxes'
```

---
title: 'Templates'
authors: ['Saeris']
description: 'CodeSandbox allows you to create your  custom templates from sandboxes'
---

## What are Templates?

> Templates are a combination of an execution [Environment]() and a set of starter files, similar to what you would find in the output of a framework CLI such as Create React App or Gatsby CLI. CodeSandbox has a number of official Templates, such as React, Vue, Gatsby and others with which you can quickly bootstrap a new project.

Reviewer: I wonder if we can simplify the explanation a bit. I think we need to drive the point that templates are nothing more than sandboxes with custom metadata to make them more easily accessible. That would make it very easy for people to understand what the concept is.

Reviewer 2: Hmm, maybe. I feel that the docs should provide more detailed technical explanations compared to say, a blog post or a tooltip. I think what you're looking for might be better off in a tooltip.

Reviewer: I think it's the correct explanation though, in my mind sandboxes keep the files and execution environment, and templates just contain a sandbox and a color. Explaining that the template keeps the execution environment might lead people to think that only templates keep it, which is not entirely true since sandboxes also keep it.

My way of explaining it would be something like:

"Templates are sandboxes that are easily accessible from your dashboard and the "Create New Sandbox" modal. They are great starting points for a new project. CodeSandbox has a number of official templates, such as React, Vue, Gatsby and others with which you can use to quickly bootstrap a new project."

Reviewer: ```suggestion
Templates are sandboxes that are easily accessible from your dashboard and the "Create New Sandbox" modal. They are great starting points for new projects. CodeSandbox has a number of official templates, such as React, Vue, Gatsby and others which you can use to quickly bootstrap a new project.
```

> Users can also create their own custom Templates, whereby they can modify an existing template with their own dependencies, files and folder structure, and configuration. Once a custom Template has been created, the user can then start new projects from the Create New Sandbox modal using one of their Templates.

Reviewer: ```suggestion
Users can also create their own custom templates, whereby they can modify an existing template with their own dependencies, files and folder structure, and configuration. Once a custom template has been created, the user can then start new projects from the "Create New Sandbox" modal using one of their templates.
```

> ## How to make a custom Template

Reviewer: ```suggestion
## How to make a custom template
```

> There are a few ways to make a Template, but the primary way to do so is from within the Project Info panel inside the Editor. When viewing one of your Sandboxes, you'll see a button at the bottom of the Project Info panel that says `Make Template`. Clicking on this will convert the Sandbox to a Template, which will freeze it and make it available inside of the Create New Sandbox modal.

Reviewer: ```suggestion
There are a few ways to make a template, but the primary way to do so is from within the "Project Info" panel inside the editor. When viewing one of your sandboxes, you'll see a button at the bottom of the "Project Info" panel that says `Make Template`. Clicking on this will convert the sandbox to a template, which will freeze it and make it available inside of the "Create New Sandbox" modal.
```

![Make Template from Sandbox](./images/templates/1.gif)

> Optionally, from the Dashboard, you can either drag and drop a Sandbox from the Overview folder, or one of your folders under My Sandboxes to the new My Templates folder. There is also an option within the Sandbox options menu to `Make Sandbox a Template`.

Reviewer: ```suggestion
Optionally, from the dashboard, you can either drag and drop a sandbox from "Overview" or one of your folders under "My Sandboxes" to the new "My Templates" folder. There is also an option within the sandbox options menu to `Make Sandbox a Template`.
```

Reviewer: It would be nice to mention that this also works for teams.

![Make Template from Dragging](./images/templates/3.gif)

![Make Template from Make Sandbox a Template](./images/templates/2.gif)

## Using Templates

> After you've made your first Template, there's a few places from which you can access and use them. First, inside of the Create New Sandbox modal, you will see a horizontal list of your Templates inside of the `Overview` tab, and all of your Templates will appear under the `My Templates` tab. Clicking on one of these will open the Template inside the Editor, where you will be prompted to fork the Sandbox as soon as you attempt to save changes.

Reviewer: ```suggestion
After you've made your first template, there's a few places from which you can access and use them. First, inside of the "Create New Sandbox" modal, you will see a horizontal list of your templates inside of the `Overview` tab, and all of your templates will appear under the `My Templates` tab. Clicking on one of these will open the template inside the editor, where you will be prompted to fork the sandbox as soon as you attempt to save changes.
```

![Sandbox Modal](./images/templates/4.gif)

> You can also find all of your Templates under the `My Templates` folder within the Dashboard. Just as with the Create New Sandbox modal, clicking on any of the Templates shown inside this folder will open a new Sandbox.

Reviewer: ```suggestion
You can also find all of your templates under the `My Templates` folder within the Dashboard. Just as with the "Create New Sandbox" modal, clicking on any of the templates shown inside this folder will open a new sandbox.
```

![Opening Template from dashboard](./images/templates/5.gif)

## Editing Templates

> Templates are frozen upon creation, preventing users from accidentally making changes to the template when they mean to use a Template to create a new Sandbox instead. In order to make changes to a Template, simply unfreeze if from the Project Info panel. Any time you try to make a change to a frozen Template, a modal will prompt you whether or not you would like to fork the Template or unfreeze it for the current session.

Reviewer: ```suggestion
Templates are frozen upon creation, preventing users from accidentally making changes to the template when they mean to use a template to create a new sandbox instead. In order to make changes to a template, simply unfreeze if from the "Project Info" panel. Any time you try to make a change to a frozen template, a modal will prompt you whether or not you would like to fork the template or unfreeze it for the current session.
```

![Fork Sandbox Modal](./images/templates/6.gif)
![unfreeze](./images/templates/7.gif)

> Any changes you make to an unfrozen Template are saved immediately. There is no need to explicitly freeze an unfrozen Template, as it will automatically be frozen again upon navigating away from the Editor or the current Sandbox session ends.

Reviewer: ```suggestion
Any changes you make to an unfrozen Template are saved immediately. There is no need to explicitly freeze an unfrozen template, as it will automatically be frozen again upon navigating away from the editor or the current sandbox session ends.
```

## Deleting Templates

> There are a few ways in which you can "delete" a Template. From the Project Info panel inside the Editor, there is a `Delete Template` button at the bottom when viewing an existing Template. Clicking this will convert the Template back to a regular Sandbox, removing it from the Create New Sandbox modal and moving back to the Overview and My Sandboxes folders inside the Dashboard.

Reviewer: ```suggestion
There are a few ways in which you can "delete" a template. From the "Project Info" panel inside the editor, there is a `Delete Template` button at the bottom when viewing an existing template. Clicking this will convert the template back to a regular sandbox, removing it from the "Create New Sandbox" modal and moving back to the "Overview" and "My Sandboxes" folders inside the dashboard.
```

![Delete template from sandbox](./images/templates/8.gif)

> Alternatively, you can use a Template's options menu inside the My Templates folder to select `Convert to Sandbox`.

Reviewer: ```suggestion
Alternatively, you can use a template's options menu inside the "My Templates" folder to select `Convert to Sandbox`.
```
